chore: Document use_effective_fields limitation for multi-shard clusters (#4557)
mongodb/terraform-provider-mongodbatlas: docs/resources/advanced_cluster.md

Reviewer: Should this guidance be added here as well?

Author: thanks, i think the cross-reference in the attribute definition is enough, since users will see it before reaching the detailed section, prefer not to duplicate it to keep things concise

* `pinned_fcv` - (Optional) Pins the Feature Compatibility Version (FCV) to the current MongoDB version with a provided expiration date. To unpin the FCV the `pinned_fcv` attribute must be removed. This operation can take several minutes as the request processes through the MongoDB data plane. Once FCV is unpinned it will not be possible to downgrade the `mongo_db_major_version`. It is advised that updates to `pinned_fcv` are done isolated from other cluster changes. If a plan contains multiple changes, the FCV change will be applied first. If FCV is unpinned past the expiration date the `pinned_fcv` attribute must be removed. The following [knowledge hub article](https://kb.corp.mongodb.com/article/000021785/) and [FCV documentation](https://www.mongodb.com/docs/atlas/tutorial/major-version-change/#manage-feature-compatibility--fcv--during-upgrades) can be referenced for more details. See [below](#pinned_fcv).
* `pit_enabled` - (Optional) Flag that indicates if the cluster uses Continuous Cloud Backup.

> * `replication_specs` - List of settings that configure your cluster regions. This attribute has one object per shard representing node configurations in each shard. For replica sets there is only one object representing node configurations. The `replication_specs` configuration for all shards within the same zone must be the same, with the exception of `instance_size` and `disk_iops` that can scale independently. Note that independent `disk_iops` values are supported for AWS Gen2 STANDARD (gp3) clusters, AWS PROVISIONED (io2) clusters, AWS HIGH_PERFORMANCE (Gen 2 io2) clusters, and Azure regions that support Extended IOPS. If this list contains more than one entry, review [Multi-shard clusters and topology changes](#multi-shard-clusters-and-topology-changes) before adding, removing, or reordering entries. See [below](#replication_specs).

Reviewer: Small edit for clarity

```suggestion
* `replication_specs` - List of settings that configure your cluster regions. This attribute has one object per shard, representing that shard's configurations. For replica sets, there is only one object representing node configurations. The `replication_specs` configuration for all shards within the same zone must be the same, with the exception of `instance_size` and `disk_iops` that can scale independently. Note that independent `disk_iops` values are supported for AWS Gen2 STANDARD (gp3) clusters, AWS PROVISIONED (io2) clusters, AWS HIGH_PERFORMANCE (Gen 2 io2) clusters, and Azure regions that support Extended IOPS. If this list contains more than one entry, review [Multi-shard clusters and topology changes](#multi-shard-clusters-and-topology-changes) before adding, removing, or reordering entries. See [below](#replication_specs).
```

Author: thanks for the suggestion but i'd like to keep this PR focused on the multi-shard limitation docs though

* `use_effective_fields` - (Optional) Controls how hardware specification fields are returned in the response. When set to true, the non-effective specs (`electable_specs`, `read_only_specs`, `analytics_specs`) fields return the hardware specifications that the client provided. When set to false (default), the non-effective specs fields show the **current** hardware specifications. Cluster auto-scaling is the primary cause for differences between initial and current hardware specifications. This opt-in feature enhances auto-scaling workflows by eliminating the need for `lifecycle.ignore_changes` blocks and preventing plan drift from Atlas-managed changes. This attribute applies to dedicated clusters, not to tenant or flex clusters. This attribute will be deprecated in provider version 2.x and removed in 3.x when the new behavior becomes default. See [Auto-Scaling with Effective Fields](#auto-scaling-with-effective-fields) for more details.

> If your cluster has more than one `replication_specs` entry, see [Multi-shard clusters and topology changes](#multi-shard-clusters-and-topology-changes) before enabling this attribute.

Reviewer 2: We are currently mentioning this reference only in effective fields, however guidance is more general to any multi-shard cluster, independent of using `use_effective_fields = true` right?

Maybe this note could be added in `cluster_type` (callout for SHARDED and GEOSHARDED with multi-shards)?

Author: good point, the guidance applies to any multi-shard cluster regardless of `use_effective_fields`. added a cross-reference in `replication_specs` rather than `cluster_type` since that's where the ordering issue manifests, let me know if you'd prefer it in `cluster_type` as well (ca25911)

More information about moving resources can be found in our [Migration Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/cluster-to-advanced-cluster-migration-guide) and in the Terraform documentation [here](https://developer.hashicorp.com/terraform/language/moved) and [here](https://developer.hashicorp.com/terraform/language/modules/develop/refactoring).

## Multi-shard clusters and topology changes

This section applies to clusters with more than one `replication_specs` entry.

**Known issue:** For clusters with more than one `replication_specs` entry, setting `use_effective_fields = true` currently has no effect. The resulting behavior is the same as leaving the attribute unset or setting it to `false`. We recommend that you do not enable this attribute for multi-shard clusters.

> Terraform matches each `replication_specs` entry to a shard by its position in the list. If you add, remove, or reorder entries, Terraform might associate an existing configuration with a different Atlas shard than intended.

Reviewer 2: Would consider placing both paragraphs as bullet points under Known issue, so it is clear there are 2 potential issues you can face (incorrect effective fields behaviour, replication_specs add/remove/reorder)

Author: good idea, restructured into two bullet points under Known issues in ca25911

More information about moving resources can be found in our [Migration Guide](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/cluster-to-advanced-cluster-migration-guide) and in the Terraform documentation [here](https://developer.hashicorp.com/terraform/language/moved) and [here](https://developer.hashicorp.com/terraform/language/modules/develop/refactoring).

## Multi-shard clusters and topology changes

> `replication_specs` is an ordered list. The provider does not expose a configurable shard identity; `external_id` is assigned by Atlas and is computed-only.

Author: addressed, the section now avoids the broad “configurable shard identity” wording and uses customer-facing wording instead: Atlas assigns shard IDs, and Terraform does not currently provide a configurable shard name that can track each shard through every topology change.
